server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java

```diff
@@ -108,4 +106,69 @@ public InputStreamContainer provideStream(int partNumber) throws IOException {
         }
 
     }
+
+    static class DecryptedReadContext<T, U> extends ReadContext {
+
+        private final U cryptoContext;
+        private final CryptoHandler<T, U> cryptoHandler;
+        private final long fileSize;
+
+        public DecryptedReadContext(ReadContext readContext, CryptoHandler<T, U> cryptoHandler) {
+            super(readContext);
+            this.cryptoHandler = cryptoHandler;
+            try {
+                this.cryptoContext = this.cryptoHandler.loadEncryptionMetadata(null);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            this.fileSize = this.cryptoHandler.estimateDecryptedLength(cryptoContext, readContext.getBlobSize());
```

Comment:
Can we avoid doing it here and delay it till getBlobSize() so that we spend compute only when getBlobSize is invoked? Also once invoked you can cache it and keep returning the cached fileSize to avoid redundant compute.